README.md
# skyline

A bar graph that displays information in real time using canvas.

[![browser support](https://ci.testling.com/tanem/skyline.png)](https://ci.testling.com/tanem/skyline)

## Live Demo

[Check out the live demo on CodePen](http://codepen.io/tanem/pen/zkanq), which uses the stand-alone files.

## Installation

Ensure [component(1)](http://component.io) is installed, then:

```sh
$ component install tanem/skyline
```

## Stand-alone

This library may be used stand-alone without the component tool. To build the stand-alone files, ensure [UglifyJS2](https://github.com/mishoo/UglifyJS2) is installed, then: 

```sh
$ make standalone
```

Then add ./standalone/skyline.css and ./standalone/skyline.js to your application and reference the `Skyline` global.

## Unit Tests

To run the unit tests, ensure [testling](https://github.com/substack/testling) is installed. Then:

```sh
$ make test
```

## API

### new Skyline

Initialize a new `Skyline`.

### Skyline#barSpacing(barSpacing:Number)

Set the bar spacing.

```js
skyline.barSpacing(3)
```

### Skyline#barWidth(barWidth:Number)

Set the bar width.

```js
skyline.barSpacing(5)
```

### Skyline#gutter(gutter:Number)

Set the gutter.

```js
skyline.gutter(50)
```

### Skyline#width(width:Number)

Set the width.

```js
skyline.width(600)
```

### Skyline#height(height:Number)

Set the height.

```js
skyline.height(300)
```

### Skyline#backgroundBarColour(backgroundBarColour:String)

Set the background bar colour.

```js
skyline.backgroundBarColour('#dedede')
```

### Skyline#historyBarColour(historyBarColour:String)

Set the history bar colour.

```js
skyline.historyBarColour('#ababab')
```

### Skyline#axesFont(axesFont:String)

Set the axes font.

```js
skyline.axesFont('14px sans-serif')
```

### Skyline#frameRate(frameRate:Number)

Set the frame rate (fps).

```js
skyline.frameRate(30)
```

### Skyline#addHit

Add a hit to the graph.

```js
skyline.addHit()
```

### Skyline#start

Start the graph animation.

```js
skyline.start()
```


CHANGELOG.md
## 3.0.8 / 2016-01-25

 * ([25463c0](https://github.com/tanem/skyline/commit/25463c0d1fb248e8defa56eca10660b7427ef9ad)) Update license (closes [#1](https://github.com/tanem/skyline/issues/1))

## v3.0.7 / 2015-03-10

 * ([ce49b19](https://github.com/tanem/skyline/commit/ce49b19e00069dc4e1212bfd509d3b85a7619f81)) Update LICENSE

## v3.0.6 / 2014-12-02

 * ([eb927f5](https://github.com/tanem/skyline/commit/eb927f580f923791c13e074923208a4e43f2d915)) Update README

## v3.0.5 / 2014-11-24

 * ([5c0a8ec](https://github.com/tanem/skyline/commit/5c0a8ec125a68b6664866f1e811e35ca4865ec84)) Update README

## v3.0.4 / 2014-10-25

 * ([f266210](https://github.com/tanem/skyline/commit/f2662100b64b6f5d1d80e875dc3f2f8c769451d8)) Tidy up README

## v3.0.3 / 2014-10-04

 * ([2394a25](https://github.com/tanem/skyline/commit/2394a25bfd0cd586e8f6ec03deca13e812c43310)) Add commit links to CHANGELOG

## v3.0.2 / 2014-10-02

 * ([1562f46](https://github.com/tanem/skyline/commit/1562f46a4408f865bf29edaac57ae31e0f57febe)) Generate CHANGELOG with mkclog

## v3.0.1 / 2014-07-07

 * ([26d4b70](https://github.com/tanem/skyline/commit/26d4b705e80a1c19f44d097135ad6ca6b2e3019f)) Add license

## v3.0.0 / 2014-05-13

 * ([1fa2cc5](https://github.com/tanem/skyline/commit/1fa2cc54b3c88dbc3b9cc1a0dc176bb670a7c57d)) Change long-winded repo name

## v2.0.25 / 2014-04-24

 * ([7bf7cdf](https://github.com/tanem/skyline/commit/7bf7cdf0b35ffd0c6c18a4b294f73fc337f0049d)) Correct component.json

## v2.0.24 / 2014-03-05

 * ([97111d4](https://github.com/tanem/skyline/commit/97111d440ced90fb961c0f8c22c1c00f6dfd62e2)) Simplified things around the html template

## v2.0.23 / 2014-03-05

 * ([2f70675](https://github.com/tanem/skyline/commit/2f70675f62c20eb4da79517896350a3444f2a62b)) Update license information

## v2.0.22 / 2014-03-05

 * ([5577559](https://github.com/tanem/skyline/commit/557755936c1af9daac4f80a2b8a7766b2700b043)) Ocd

## v2.0.21 / 2014-03-01

 * ([e7789b8](https://github.com/tanem/skyline/commit/e7789b8e7532dab9b51bb3629f8858d20be5a96d)) Simplify standalone build

## v2.0.20 / 2014-03-01

 * ([966e3c5](https://github.com/tanem/skyline/commit/966e3c5a5e72c8534672c757cc96ef731d0068bb)) Run unit tests with testling

## v2.0.19 / 2014-02-13

 * ([6dd552b](https://github.com/tanem/skyline/commit/6dd552bbd592bd9ebe847074ef53f1c02993ace9)) Update unit tests to use Sinon instead of Spooks
 * ([1f855c9](https://github.com/tanem/skyline/commit/1f855c93127d9103c9afdac66298b4de2687946c)) Update JSHint rules

## v2.0.18 / 2014-01-24

 * ([a1a625c](https://github.com/tanem/skyline/commit/a1a625c7b832de5566abe232800cae36e7e50ed3)) Tweak background bar colour.

## v2.0.17 / 2014-01-24

 * ([b4cc7b3](https://github.com/tanem/skyline/commit/b4cc7b33f2683a03cece3b813ee156d0c0b42f4b)) Build new standalone CSS file.

## v2.0.16 / 2014-01-24

 * ([8db27b6](https://github.com/tanem/skyline/commit/8db27b6ab7d5b4fa9f64caf9363dcc5b289f62ed)) Update CSS and demo.

## v2.0.15 / 2014-01-24

 * ([eb1040e](https://github.com/tanem/skyline/commit/eb1040ef311660e9d768b3597711d329c92e7ad5)) Decruft CSS.

## v2.0.14 / 2014-01-24

 * ([796cec4](https://github.com/tanem/skyline/commit/796cec4c3082277e6417721f25e1434e2a1c905a)) Update component.json.

## v2.0.13 / 2014-01-24

 * ([57a7a3f](https://github.com/tanem/skyline/commit/57a7a3f58b1fb15ba67c80b264bf8148336a9382)) Clean up Makefile.

## v2.0.12 / 2014-01-22

 * ([678f601](https://github.com/tanem/skyline/commit/678f601e3dad9fdd05d2ae4a8c34d8fdc09f9318)) Remove stats target from Makefile.

## v2.0.11 / 2014-01-22

 * ([444e865](https://github.com/tanem/skyline/commit/444e8651d3e10cc06252ecb919cd22b5b7fdf89a)) Update README.

## v2.0.10 / 2014-01-22

 * ([1244b3f](https://github.com/tanem/skyline/commit/1244b3fb69f23301decdf5f82bb91e6daee91315)) Update README.

## v2.0.9 / 2014-01-22

 * ([15bad18](https://github.com/tanem/skyline/commit/15bad1870058de1fefe7301e8f886263d2fb196b)) Update README.

## v2.0.8 / 2014-01-22

 * ([0c96d6b](https://github.com/tanem/skyline/commit/0c96d6b42c8ad54cf3476d9f29d61bbf36161336)) Update README.

## v2.0.7 / 2014-01-22

 * ([73e8d20](https://github.com/tanem/skyline/commit/73e8d20ff0a6f6472491d502485cdca1d64264e6)) Update README.

## v2.0.6 / 2014-01-22

 * ([6b9630e](https://github.com/tanem/skyline/commit/6b9630ebbfd1368873e60a7b8d9a1efa1922f46d)) Correct version in component.json.

## v2.0.5 / 2014-01-22

 * ([4a9fd2a](https://github.com/tanem/skyline/commit/4a9fd2ab68877dfe5339d4e85598af9dff0f72a9)) Add global NPM dependencies to README.

## v2.0.4 / 2014-01-20

 * ([6701c1d](https://github.com/tanem/skyline/commit/6701c1db99e6710ae984fc2cd45f0c0f526101df)) Update README with CodePen demo link.

## v2.0.3 / 2014-01-20

 * ([bb28124](https://github.com/tanem/skyline/commit/bb281245ff570a4855dcd7eeb72b56d7e162dcba)) Remove build dir from git index.

## v2.0.2 / 2014-01-20

 * ([ab18271](https://github.com/tanem/skyline/commit/ab182711507ad4636ccf5294f57df07d59c1fc61)) Remove components dir from git index.

## v2.0.1 / 2014-01-20

 * ([53185e6](https://github.com/tanem/skyline/commit/53185e67757ed96f917e8223b09a61dcf3b7f1dd)) Update .gitignore.

## v2.0.0 / 2014-01-20

 * ([6b9c601](https://github.com/tanem/skyline/commit/6b9c6012d85a5f24c204883580496109c3d17161)) Rewritten as a component.

## v1.0.2 / 2013-12-13

 * ([15ceb87](https://github.com/tanem/skyline/commit/15ceb871f92e0507459088e6b50f975e2694c9bd)) README updated.

## v1.0.1 / 2013-12-13

 * ([b59f14b](https://github.com/tanem/skyline/commit/b59f14b8cab673187e8eaf3a20cd259c500572e7)) Updates for new repo name.
 * ([d9219cd](https://github.com/tanem/skyline/commit/d9219cdda7cf5345fc6ee26b1bfccaf15029f224)) Updated for new repo name.
 * ([952cf6f](https://github.com/tanem/skyline/commit/952cf6f21d971ffb5b4e82a6ecc6ff3277a451aa)) Updated demo hits adder to start with reqAnimFrame.

standalone/skyline.js

;(function(){

/**
 * Require the module at `name`.
 *
 * @param {String} name
 * @return {Object} exports
 * @api public
 */

function require(name) {
  var module = require.modules[name];
  if (!module) throw new Error('failed to require "' + name + '"');

  if (!('exports' in module) && typeof module.definition === 'function') {
    module.client = module.component = true;
    module.definition.call(this, module.exports = {}, module);
    delete module.definition;
  }

  return module.exports;
}

/**
 * Registered modules.
 */

require.modules = {};

/**
 * Register module at `name` with callback `definition`.
 *
 * @param {String} name
 * @param {Function} definition
 * @api private
 */

require.register = function (name, definition) {
  require.modules[name] = {
    definition: definition
  };
};

/**
 * Define a module's exports immediately with `exports`.
 *
 * @param {String} name
 * @param {Generic} exports
 * @api private
 */

require.define = function (name, exports) {
  require.modules[name] = {
    exports: exports
  };
};
require.register("component~raf@1.1.3", function (exports, module) {
/**
 * Expose `requestAnimationFrame()`.
 */

exports = module.exports = window.requestAnimationFrame
  || window.webkitRequestAnimationFrame
  || window.mozRequestAnimationFrame
  || window.oRequestAnimationFrame
  || window.msRequestAnimationFrame
  || fallback;

/**
 * Fallback implementation.
 */

var prev = new Date().getTime();
function fallback(fn) {
  var curr = new Date().getTime();
  var ms = Math.max(0, 16 - (curr - prev));
  var req = setTimeout(fn, ms);
  prev = curr;
  return req;
}

/**
 * Cancel.
 */

var cancel = window.cancelAnimationFrame
  || window.webkitCancelAnimationFrame
  || window.mozCancelAnimationFrame
  || window.oCancelAnimationFrame
  || window.msCancelAnimationFrame
  || window.clearTimeout;

exports.cancel = function(id){
  cancel.call(window, id);
};

});

require.register("kenany~isinteger@1.0.4", function (exports, module) {
/**
 * Check if a Number is an integer
 *
 * @param {Number} x
 * @return {Boolean} is integer
 * @api public
 */
module.exports = function(x) {
  return (x === Math.round(x));
};
});

require.register("component~bind@0.0.1", function (exports, module) {

/**
 * Slice reference.
 */

var slice = [].slice;

/**
 * Bind `obj` to `fn`.
 *
 * @param {Object} obj
 * @param {Function|String} fn or string
 * @return {Function}
 * @api public
 */

module.exports = function(obj, fn){
  if ('string' == typeof fn) fn = obj[fn];
  if ('function' != typeof fn) throw new Error('bind() requires a function');
  var args = [].slice.call(arguments, 2);
  return function(){
    return fn.apply(obj, args.concat(slice.call(arguments)));
  }
};

});

require.register("component~domify@1.2.2", function (exports, module) {

/**
 * Expose `parse`.
 */

module.exports = parse;

/**
 * Wrap map from jquery.
 */

var map = {
  legend: [1, '<fieldset>', '</fieldset>'],
  tr: [2, '<table><tbody>', '</tbody></table>'],
  col: [2, '<table><tbody></tbody><colgroup>', '</colgroup></table>'],
  _default: [0, '', '']
};

map.td =
map.th = [3, '<table><tbody><tr>', '</tr></tbody></table>'];

map.option =
map.optgroup = [1, '<select multiple="multiple">', '</select>'];

map.thead =
map.tbody =
map.colgroup =
map.caption =
map.tfoot = [1, '<table>', '</table>'];

map.text =
map.circle =
map.ellipse =
map.line =
map.path =
map.polygon =
map.polyline =
map.rect = [1, '<svg xmlns="http://www.w3.org/2000/svg" version="1.1">','</svg>'];

/**
 * Parse `html` and return the children.
 *
 * @param {String} html
 * @return {Array}
 * @api private
 */

function parse(html) {
  if ('string' != typeof html) throw new TypeError('String expected');
  
  // tag name
  var m = /<([\w:]+)/.exec(html);
  if (!m) return document.createTextNode(html);

  html = html.replace(/^\s+|\s+$/g, ''); // Remove leading/trailing whitespace

  var tag = m[1];

  // body support
  if (tag == 'body') {
    var el = document.createElement('html');
    el.innerHTML = html;
    return el.removeChild(el.lastChild);
  }

  // wrap map
  var wrap = map[tag] || map._default;
  var depth = wrap[0];
  var prefix = wrap[1];
  var suffix = wrap[2];
  var el = document.createElement('div');
  el.innerHTML = prefix + html + suffix;
  while (depth--) el = el.lastChild;

  // one element
  if (el.firstChild == el.lastChild) {
    return el.removeChild(el.firstChild);
  }

  // several elements
  var fragment = document.createDocumentFragment();
  while (el.firstChild) {
    fragment.appendChild(el.removeChild(el.firstChild));
  }

  return fragment;
}

});

require.register("component~autoscale-canvas@0.0.3", function (exports, module) {

/**
 * Retina-enable the given `canvas`.
 *
 * @param {Canvas} canvas
 * @return {Canvas}
 * @api public
 */

module.exports = function(canvas){
  var ctx = canvas.getContext('2d');
  var ratio = window.devicePixelRatio || 1;
  if (1 != ratio) {
    canvas.style.width = canvas.width + 'px';
    canvas.style.height = canvas.height + 'px';
    canvas.width *= ratio;
    canvas.height *= ratio;
    ctx.scale(ratio, ratio);
  }
  return canvas;
};
});

require.register("skyline", function (exports, module) {
'use strict';

var raf = require("component~raf@1.1.3");
var bind = require("component~bind@0.0.1");
var domify = require("component~domify@1.2.2");
var template = require("skyline/template.html");
var autoscale = require("component~autoscale-canvas@0.0.3");

/**
 * Expose `Skyline`.
 */

module.exports = Skyline;

/**
 * Initialize a new `Skyline`.
 *
 * @api public
 */

function Skyline() {
  this._barSpacing = 2;
  this._barWidth = 4;
  this._gutter = 40;
  this._width = 780;
  this._height = 280;
  this._backgroundBarColour = '#ddd';
  this._historyBarColour = '#00cccc';
  this._axesFont = '13px sans-serif';
  this._frameRate = 20;
  this.buffer = 0;
  this.el = domify(template);
  this.animate = bind(this, 'animate');
  this.backgroundBarCb = bind(this, 'backgroundBarCb');
  this.historyBarCb = bind(this, 'historyBarCb');
}

/**
 * Set the bar spacing.
 *
 * ```js
 * skyline.barSpacing(3)
 * ```
 *
 * @param {Number} barSpacing
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.barSpacing = function(barSpacing){
  this._barSpacing = barSpacing;
  return this;
};

/**
 * Set the bar width.
 *
 * ```js
 * skyline.barWidth(5)
 * ```
 *
 * @param {Number} barWidth
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.barWidth = function(barWidth){
  this._barWidth = barWidth;
  return this;
};

/**
 * Set the gutter.
 *
 * ```js
 * skyline.gutter(50)
 * ```
 *
 * @param {Number} gutter
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.gutter = function(gutter){
  this._gutter = gutter;
  return this;
};

/**
 * Set the width.
 *
 * ```js
 * skyline.width(600)
 * ```
 *
 * @param {Number} width
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.width = function(width){
  this._width = width;
  return this;
};

/**
 * Set the height.
 *
 * ```js
 * skyline.height(300)
 * ```
 *
 * @param {Number} height
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.height = function(height){
  this._height = height;
  return this;
};

/**
 * Set the background bar colour.
 *
 * ```js
 * skyline.backgroundBarColour('#dedede')
 * ```
 *
 * @param {String} backgroundBarColour
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.backgroundBarColour = function(backgroundBarColour){
  this._backgroundBarColour = backgroundBarColour;
  return this;
};

/**
 * Set the history bar colour.
 *
 * ```js
 * skyline.historyBarColour('#ababab')
 * ```
 *
 * @param {String} historyBarColour
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.historyBarColour = function(historyBarColour){
  this._historyBarColour = historyBarColour;
  return this;
};

/**
 * Set the axes font.
 *
 * ```js
 * skyline.axesFont('14px sans-serif')
 * ```
 *
 * @param {String} axesFont
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.axesFont = function(axesFont){
  this._axesFont = axesFont;
  return this;
};

/**
 * Set the frame rate (fps).
 *
 * ```js
 * skyline.frameRate(30)
 * ```
 *
 * @param {Number} frameRate
 * @return {Skyline} self
 * @api public
 */

Skyline.prototype.frameRate = function(frameRate){
  this._frameRate = frameRate;
  return this;
};

/**
 * Create the history array.
 *
 * @api private
 */

 Skyline.prototype.createHistoryArray = function(){
  this.history = [];
  var i = 0, j = this._width / (this._barSpacing + this._barWidth);
  for (; i < j; i++) this.history[i] = 0;
};

/**
 * Draw the background bars.
 *
 * @api private
 */

Skyline.prototype.drawBackground = function(){
  this.drawBars(this.backgroundBarCb);
};

/**
 * Draw a single background bar at the required position.
 *
 * @param {Number} xPos
 * @api private
 */

Skyline.prototype.backgroundBarCb = function(xPos){
  this.bgCtx.fillRect(xPos, this._gutter, this._barWidth, this._height - this._gutter * 2);
};

/**
 * Initialise the background canvas properties.
 *
 * @api private
 */

Skyline.prototype.initialiseBackgroundCanvas = function(){
  var canvas = this.el.querySelector('.skyline-background');
  canvas.width = this._width;
  canvas.height = this._height;
  autoscale(canvas);
  this.bgCtx = canvas.getContext('2d');
  this.bgCtx.fillStyle = this._backgroundBarColour;
};

/**
 * Initialise the non-changing history canvas properties.
 *
 * @api private
 */

Skyline.prototype.initialiseHistoryCanvas = function(){
  var canvas = this.el.querySelector('.skyline-history');
  canvas.width = this._width;
  canvas.height = this._height;
  autoscale(canvas);
  this.historyCtx = canvas.getContext('2d');
  this.historyCtx.fillStyle = this._historyBarColour;
};

/**
 * Start the graph animation.
 *
 * @api public
 */

Skyline.prototype.start = function(){
  this.createHistoryArray();
  this.initialiseBackgroundCanvas();
  this.initialiseHistoryCanvas();
  this.drawBackground();
  this.animateInterval = 1000 / this._frameRate;
  this.rafId = raf(this.animate);
};

/**
 * Animation loop.
 *
 * @api private
 */

Skyline.prototype.animate = function(){
  this.rafId = raf(this.animate);
  var now = Date.now();
  this.animateStart = this.animateStart || now;
  var animateDelta = now - this.animateStart;
  
  // Process the history interval at the required framerate.
  if (animateDelta > this.animateInterval) {

    // See: http://codetheory.in/controlling-the-frame-rate-with-requestanimationframe/
    // Basically, we want to eliminate the creep which may be present after
    // animateDelta > interval, in order to keep the framerate as accurate as possible.
    this.animateStart = now - (animateDelta % this.animateInterval);
    this.processHistoryFrame();
  }
};

/**
 * Process a history frame - draw the bar graph at this point in time.
 *
 * @api private
 */

Skyline.prototype.processHistoryFrame = function(){
  for (var i = 0, j = this.history.length - 1; i < j; i++) {
    this.history[i] = this.history[i + 1];
  }
  this.history[this.history.length - 1] = this.buffer;
  this.drawHistory();
  this.buffer = 0;
};

/**
 * Draw the history bars.
 *
 * @api private
 */

Skyline.prototype.drawHistory = function(){
  var maxHits = Math.max.apply(null, this.history);
  if (!this.maxAxesPoint || maxHits >= this.maxAxesPoint) this.drawAxes(maxHits);
  this.historyMultiplier = (this._height - this._gutter * 2) / this.maxAxesPoint;
  this.historyCtx.clearRect(0, 0, this._width, this._height);
  this.historyStartIndex = this.history.length - 1;
  this.drawBars(this.historyBarCb);
};

/**
 * Draw a single history bar scaled to the correct height.
 *
 * @param {Number} xPos
 * @api private
 */

Skyline.prototype.historyBarCb = function(xPos){
  var hits = this.history[this.historyStartIndex--] * this.historyMultiplier;
  this.historyCtx.fillRect(xPos, this._height - this._gutter - hits, this._barWidth, hits);
};

/**
 * Draw graph bars by executing a callback at the required interval.
 *
 * @param {Function} cb
 * @api private
 */

Skyline.prototype.drawBars = function(cb){
  var xPos = this._width - this._gutter - this._barWidth, start = this._gutter;
  for (; xPos >= start; xPos -= (this._barSpacing + this._barWidth)) cb(xPos);
};

/**
 * Draw the left and right y-axes scaled to `maxHits`.
 *
 * @param {Number} maxHits
 * @api private
 */

Skyline.prototype.drawAxes = function(maxHits){
  maxHits = maxHits || 1;
  var points = [0, maxHits, maxHits * 2, maxHits * 3, maxHits * 4];
  this.maxAxesPoint = points[4];
  this.bgCtx.font = this._axesFont;
  this.drawLeftYAxis(points);
  this.drawRightYAxis(points);
};

/**
 * Draw the left y-axis using the given points.
 *
 * @param {Array} points
 * @api private
 */

Skyline.prototype.drawLeftYAxis = function(points){
  var xPos = this._gutter - 5;
  var drawingAreaHeight = this._height - this._gutter * 2;
  this.bgCtx.clearRect(0, 0, xPos, this._height);
  this.bgCtx.textAlign = 'right';
  this.bgCtx.textBaseline = 'middle';
  this.bgCtx.fillText(points[4], xPos, this._gutter);
  this.bgCtx.fillText(points[3], xPos, this._gutter + drawingAreaHeight / 4);
  this.bgCtx.fillText(points[2], xPos, this._gutter + drawingAreaHeight / 2);
  this.bgCtx.fillText(points[1], xPos, this._height - this._gutter - drawingAreaHeight / 4);
  this.bgCtx.fillText(points[0], xPos, this._height - this._gutter);
};

/**
 * Draw the right y-axis using the given points.
 *
 * @param {Array} points
 * @api private
 */

Skyline.prototype.drawRightYAxis = function(points){
  var xPos = this._width - this._gutter + 5;
  var drawingAreaHeight = this._height - this._gutter * 2;
  this.bgCtx.clearRect(xPos, 0, this._width - xPos, this._height);
  this.bgCtx.textAlign = 'left';
  this.bgCtx.textBaseline = 'middle';
  this.bgCtx.fillText(points[4], xPos, this._gutter);
  this.bgCtx.fillText(points[3], xPos, this._gutter + drawingAreaHeight / 4);
  this.bgCtx.fillText(points[2], xPos, this._gutter + drawingAreaHeight / 2);
  this.bgCtx.fillText(points[1], xPos, this._height - this._gutter - drawingAreaHeight / 4);
  this.bgCtx.fillText(points[0], xPos, this._height - this._gutter);
};

/**
 * Add a hit to the graph.
 *
 * @api public
 */

Skyline.prototype.addHit = function(){
  this.buffer++;
};
});

require.define("skyline/template.html", "<div class=\"skyline\">\n  <canvas class=\"skyline-background\"></canvas>\n  <canvas class=\"skyline-history\"></canvas>\n</div>");

if (typeof exports == "object") {
  module.exports = require("skyline");
} else if (typeof define == "function" && define.amd) {
  define([], function(){ return require("skyline"); });
} else {
  this["Skyline"] = require("skyline");
}
})()
